README.md

```diff
@@ -1,5 +1,340 @@
+            'name' => 'example-name',
+        ]);
+    }
+}
+```
+
+### Dispatching a query
+
+As with [dispatching a command][6], we can use the batteries-included query bus which can be included using the `QueryBusTrait`.
+
+```php
+namespace My\Query;
+
+use Zumba\CQRS\QueryBusTrait;
+
+class Example {
+    use QueryBusTrait;
+
+    public function query(): array
+    {
+        $response = $this->queryBus()->dispatch(new MyQuery([
+            'id' => 'example-id',
+        ]));
+        return [
+            'id' => $response['id'],
+            'name' => $response['name'],
+        ];
+    }
+
+}
+
+var_dump((new Example)->query());
+// [
+//     'id' => 'example-id',
+//     'name' => 'example-name',
+// ]
+```
+
+## Providers
+
+Providers offer the ability for the command bus to be able to supply a handler for a DTO dispatched to the bus.
+
+`zumba\cqrs` provides several providers out of the box.
+
+### Class Provider
+
+The `Zumba\CQRS\Provider\ClassProvider` can be used to construct handlers for a DTO via a handler factory.
+This can be used when the handlers have complex dependencies (ie dependencies that require configuration).
+
+To utilize this, create a handler factory in the same namespace as the DTO. Assume our DTO from the [create a command][4] section
+of `MyCommand`, the handler factory would look like this:
+
+```php
+<?php
+
+namespace My\Command;
+
+use Zumba\CQRS\Command\HandlerFactory;
+use Zumba\CQRS\Command\Handler;
+
+class MyCommandHandlerFactory implements HandlerFactory
+{
+    public static function make(): Handler
+    {
+        return new MyCommandHandler();
+    }
+}
+```
+
+### Method Provider
+
+The `Zumba\CQRS\Provider\MethodProvider` is similar to the `ClassProvider` except the handler itself can serve as its own factory.
+
+The `MyCommandHandler` can be adapted from the [create a command handler][5] section:
+
+```php
+<?php
+
+namespace My\Command;
+
+use Zumba\CQRS\Command\Command;
+use Zumba\CQRS\Command\CommandResponse;
+use Zumba\CQRS\Command\Handler;
+use Zumba\CQRS\Command\HandlerFactory;
+use Zumba\CQRS\CommandService;
+
+final class MyCommandHandler implements Handler, HandlerFactory
+{
+    public function handle(Command $command, CommandService $commandService): CommandResponse
+    {
+        echo $command->message;
+        return CommandResponse::fromSuccess();
+    }
+
+    public static function make(): Handler
+    {
+        return new static();
+    }
+}
+```
+
+### Simple Dependency Provider
+
+If cases where all dependencies can be instantiated without parameters, the `Zumba\CQRS\Provider\SimpleDependencyProvider` can be used to construct the handler without the need of a `HandlerFactory`.
+
+### Create a custom provider
+
+To create your own custom provider, implement the `Zumba\CQRS\Provider` interface and create the bus with the custom provider.
+You can also provide additional providers in which to fallback on if your CustomProvider can't accommodate the DTO.
```

Comment:
I think we should remove `in which` from this sentence, reads a little awkwardly...